src/session-user/session-user.service.ts

```diff
@@ -1,4 +1,119 @@
+    const userCompletedSession = userCourse.course[0].session;
+
+    const userSessionIds = userCompletedSession.map((session) => {
+      if (session.completed) return session.id;
+    });
+
+    const courseSessionIds = courseSession.map((session) => {
+      if (session.status === STORYBLOK_STORY_STATUS_ENUM.PUBLISHED) return session.id;
+    });
+
+    return _.xor(courseSessionIds, userSessionIds).length == 0;
+  }
+
+  public async createSessionUser(
+    { uid }: IFirebaseUser,
+    { sessionId }: CreateSessionUserDto,
+  ): Promise<SessionUserEntity> {
+    const user = await this.userRepository.findOne({ firebaseUid: uid });
+
+    const { courseId } = await this.sessionService.getSession(sessionId);
+
+    const courseSessions = await this.courseService.getCourseSessions(courseId);
+
+    if (!courseSessions) {
+      throw new HttpException('COURSE SESSIONS NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    let courseUser = await this.courseUserService.getCourseUser({ userId: user.id, courseId });
+
+    if (!courseUser) {
+      courseUser = await this.courseUserService.createCourseUser({ userId: user.id, courseId });
+    }
+
+    const createSessionUserObject = this.sessionUserRepository.create({
+      sessionId,
+      courseUserId: courseUser.id,
+      completed: false,
+    });
+
+    let sessionUser = await this.sessionUserRepository.findOne({
+      courseUserId: courseUser.id,
+      sessionId,
+    });
+
+    if (!sessionUser) {
+      sessionUser = await this.sessionUserRepository.save(createSessionUserObject);
+    }
+
+    return sessionUser;
+  }
+
+  public async updateSessionUser({ uid }: IFirebaseUser, sessionId: string) {
+    const user = await this.userRepository.findOne({ firebaseUid: uid });
+
+    const { courseId } = await this.sessionService.getSession(sessionId);
+
+    if (!courseId) {
+      throw new HttpException('COURSE NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    const courseSessions = await this.courseService.getCourseSessions(courseId);
+
+    const courseUser = await this.courseUserService.getCourseUser({ userId: user.id, courseId });
+
+    if (!courseUser) {
+      throw new HttpException('COURSE USER NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    const sessionUser = await this.sessionUserRepository.findOne({
+      where: { courseUserId: courseUser.id, sessionId },
+    });
+
+    if (!sessionUser) {
+      throw new HttpException('SESSION USER NOT FOUND', HttpStatus.NOT_FOUND);
+    }
+
+    sessionUser.completed = true;
+
+    await this.sessionUserRepository.save(sessionUser);
+
+    const userObject = await this.userService.getUser(user);
+
+    const markCourseComplete = this.markCourseComplete(userObject, courseSessions.session);
+    if (markCourseComplete) {
+      await this.courseUserService.completeCourse({ userId: user.id, courseId });
+      userObject.course[0].completed = true;
+    }
+
+    return userObject;
```

Comment:
i think we should return just the `Session` here, but we'd also like to know if the course is completed, so maybe an extra field `courseComplete` returned, or return the `Course`. but we dont need the full `getUser` data here